pkg/collector/scheduler/job.go

```diff
@@ -10,97 +10,187 @@ import (
 	"sync"
 	"time"
 
-	"github.com/DataDog/datadog-agent/pkg/util/log"
-
 	"github.com/DataDog/datadog-agent/pkg/collector/check"
 	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+type jobBucket struct {
+	jobs []check.Check
+	mu   sync.RWMutex // to protect critical sections in struct's fields
+}
+
+func (jb *jobBucket) size() int {
+	jb.mu.RLock()
+	defer jb.mu.RUnlock()
+
+	return len(jb.jobs)
+}
+
+func (jb *jobBucket) addJob(c check.Check) {
+	jb.mu.Lock()
+	defer jb.mu.Unlock()
+
+	jb.jobs = append(jb.jobs, c)
+}
+
 // jobQueue contains a list of checks (called jobs) that need to be
 // scheduled at a certain interval.
 type jobQueue struct {
-	interval time.Duration
-	stop     chan bool // to stop this queue
-	stopped  chan bool // signals that this queue has stopped
-	ticker   *time.Ticker
-	jobs     []check.Check
-	running  bool
-	health   *health.Handle
-	mu       sync.RWMutex // to protect critical sections in struct's fields
+	interval            time.Duration
+	stop                chan bool // to stop this queue
+	stopped             chan bool // signals that this queue has stopped
+	buckets             []*jobBucket
+	bucketTicker        *time.Ticker
+	lastTick            time.Time
+	sparseStep          uint
+	currentBucketIdx    uint
+	schedulingBucketIdx uint
+	running             bool
+	health              *health.Handle
+	mu                  sync.RWMutex // to protect critical sections in struct's fields
 }
 
 // newJobQueue creates a new jobQueue instance
 func newJobQueue(interval time.Duration) *jobQueue {
-	return &jobQueue{
-		interval: interval,
-		ticker:   time.NewTicker(time.Duration(interval)),
-		stop:     make(chan bool),
-		stopped:  make(chan bool),
-		health:   health.Register("collector-queue"),
+	jq := &jobQueue{
+		interval:     interval,
+		stop:         make(chan bool),
+		stopped:      make(chan bool),
+		health:       health.Register("collector-queue"),
+		bucketTicker: time.NewTicker(time.Second),
+	}
+
+	var nb int
+	if interval <= time.Second {
+		nb = 1
+	} else {
+		nb = int(interval.Truncate(time.Second).Seconds())
+	}
+	for i := 0; i < nb; i++ {
+		bucket := &jobBucket{}
+		jq.buckets = append(jq.buckets, bucket)
+	}
+
+	// compute step for sparse scheduling
+	switch nb % 2 {
```

Comment:
First time I see this algorithm but it should be a good addition.
I think we have an unhandled edge case with `2` : https://play.golang.org/p/8aPXsdpVxzY
Do you have examples of other implementations of thins in case we missed anything else ?

Reply:
I didn't base myself off any other implementation. I deduced this from modular arithmetic. The goals were:
1) let's schedule the checks as far apart from each other as possible, so as to smoothen the CPU load over the entire interval (best effort).
2) make sure that the steps would make us fill every bucket with a single check, before adding the next round of checks. 

If you need formal mathematical proof I can work on it. But basically the above will hold true (we could improve and make it even more sparse, but I don't think we need to go crazy) when:
a) Even number of buckets: we use the first odd index to the left of the middle index (doesn't matter if that index is even or odd - ie. 20 -> 9, 22 -> 9). 
b) Odd number of buckets: we use the middle index. (ie. 9 -> 4, 15 -> 7).

The reason this works is because when we wrap around the modular ring, we "land" in a bucket that was previously unused. With case (b) this is quite easy to see, we are always one off the previous location when wrapping around. (ie. 9: 0, 4, 8, 3, 7, 2, 6, 1, 5). With case (a) it's less trivial, you need to rely on the fact that since the number of buckets is even, and our "step" is odd, in the general case we will have three "steps" before wrapping around and since `odd*odd = odd`, so by definition we will not land on the same bucket, and the offset is such that we will continue to wrap around landing in free buckets until all buckets have been filled.

Note:  for (a) if N <= 6, the general case begins to break down, as we go down to a step of just 1, but that will work just fine as well. It's easy to see that steps of 2 or 3 would not work as we'd always be hitting the same buckets.